README.md
# koa-http-server

[![License: MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](https://opensource.org/licenses/MIT) [![codecov](https://codecov.io/gh/pkuosa-gabriel/koa-http-server/branch/master/graph/badge.svg)](https://codecov.io/gh/pkuosa-gabriel/koa-http-server) [![codebeat badge](https://codebeat.co/badges/76d490f7-1f76-4e04-8582-4b32e0fb362f)](https://codebeat.co/projects/github-com-pkuosa-gabriel-koa-http-server-master)

A simple http server based on [koa.js](https://github.com/koajs/koa). A toy version is deployed on [Heroku](https://koa-http-server.herokuapp.com), feel free to have a try.

Notes can be found [here](https://github.com/pkuosa-gabriel/notes/).

If you have any questions or suggestions, just send me an [e-mail](mailto:wuzihua@pku.edu.cn). If you find any bugs, please create an issue in this repository. Pull requests are also welcome.

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [0.0.4] - 2018-10-17

### Changed

- Beautify the home page.

## [0.0.3] - 2018-10-17

### Changed

- File size is limited to 5 kB.
- Null upload is rejected.
- Add a little bit CSS.

## [0.0.2] - 2018-10-16

### Added

- Users can submit a text at `/` via `POST /upload/text`.
- Users can upload a file at `/` via `POST /upload/file`.

## [0.0.1] - 2018-10-15

### Added

- Basic http-server based on [koa.js](https://github.com/koajs/koa).
- Basic test based on [mocha](https://github.com/mochajs/mocha), [chai](https://github.com/chaijs/chai) and [chai-http](https://github.com/chaijs/chai-http).
- Codebeat quality assesment.
- Codecov report.
- Mergify for auto-merge.
- Renovate for dependency update.
- Travis CI.
- Continuous deployment to Heroku.


test/index.spec.js
const {readFileSync} = require('fs');
const debug = require('debug')('test:index');
const chai = require('chai');
const chaiHttp = require('chai-http');
const chaiString = require('chai-string');
const {server} = require('../index');
const expect = chai.expect;

chai.use(chaiHttp).use(chaiString);

describe('Basic routes', () => {
  after(() => {
    server.close();
    debug('Server has been closed.');
  });

  it('should get HOME', (done) => {
    chai.request(server)
      .get('/')
      .end((err, res) => {
        expect(res).to.have.status(200);
        expect(res).to.be.html;
        expect(res.text).to.containIgnoreSpaces('A simple http server powered by koa.js');
        done();
      });
  });

  it('should get ABOUT', (done) => {
    chai.request(server)
      .get('/about')
      .end((err, res) => {
        expect(res).to.have.status(200);
        expect(res.text).equal('This is a simple http-server powered by koa.js');
        done();
      });
  });

  it('should redirect /github to GITHUB', (done) => {
    chai.request(server)
      .get('/github').redirects(0)
      .end((err, res) => {
        expect(res).to.have.status(302);
        expect(res).to.redirectTo('https://github.com/pkuosa-gabriel/koa-http-server');
        done();
      });
  });

  it('should handle NOT FOUND', (done) => {
    chai.request(server)
      .get('/random-route')
      .end((err, res) => {
        expect(res).to.have.status(404);
        expect(res.text).equal('Maybe you have entered a wrong path');
        done();
      });
  });

  it('should upload a text', (done) => {
    chai.request(server)
      .post('/upload/text')
      .set('content-type', 'application/json')
      .send({'textLayout': 'hello'})
      .end((err, res) => {
        expect(res).to.have.status(200);
        expect(res.text).equal('You\'ve sent the text: hello');
        done();
      });
  });

  it('should upload a file', (done) => {
    chai.request(server)
      .post('/upload/file')
      .attach('upload', readFileSync('./test/test.txt'), 'test.txt')
      .end((err, res) => {
        expect(res).to.have.status(200);
        expect(res.text).equal('You\'ve uploaded the file: test.txt');
        done();
      });
  });

  it('should reject null upload', (done) => {
    chai.request(server)
      .post('/upload/file')
      .end((err, res) => {
        expect(res).to.have.status(406);
        expect(res.text).equal('You must upload a file');
        done();
      });
  });

  it('should reject a big file', (done) => {
    chai.request(server)
      .post('/upload/file')
      .attach('upload', readFileSync('./test/big.jpg'), 'big.jpg')
      .end((err, res) => {
        expect(res).to.have.status(406);
        expect(res.text).equal('The file is too large');
        done();
      });
  });
});


router.js
const debug = require('debug')('server:router');
const fs = require('fs');
const path = require('path');
const uuid = require('uuid');
const router = require('koa-router')();

const FILE_SIZE_LIMIT = 5 * 1024;
const route = router.use('*', recv)
  .get('home', '/', home)
  .get('about', '/about', about)
  .get('github', '/github', github)
  .get('not-found', '*', fallback)
  .post('upload-text', '/upload/text', uploadText)
  .post('upload-file', '/upload/file', uploadFile);

/**
 * Before routing
 * @param {*} ctx
 * @param {*} next
 */
async function recv(ctx, next) {
  debug(`Request for ${ctx.path} received.`);
  await next();
}

/**
 * Handle home
 * @param {*} ctx
 */
async function home(ctx) {
  await ctx.render('index');
}

/**
 * Handle about
 * @param {*} ctx
 */
async function about(ctx) {
  ctx.body = 'This is a simple http-server powered by koa.js';
}


/**
 * Redirect to github
 * @param {*} ctx
 */
async function github(ctx) {
  ctx.redirect('https://github.com/pkuosa-gabriel/koa-http-server');
}

/**
 * Fallback for routes not found
 * @param {*} ctx
 */
async function fallback(ctx) {
  debug(`No route found for ${ctx.path}`);
  ctx.status = 404;
  ctx.body = 'Maybe you have entered a wrong path';
}

/**
 * Handle text upload
 * @param {*} ctx
 */
async function uploadText(ctx) {
  const text = ctx.request.body.textLayout;
  debug(`Receive text: ${text}`);
  ctx.body = `You've sent the text: ${text}`;
}

/**
 * Handle file upload
 * @param {*} ctx
 */
async function uploadFile(ctx) {
  if (ctx.request.files === undefined) {
    ctx.status = 406;
    ctx.body = 'You must upload a file';
    return;
  }
  const file = ctx.request.files.upload;
  const reader = fs.createReadStream(file.path);
  const fileName = file.name;
  const fileSize = file.size;
  if (fileSize > FILE_SIZE_LIMIT) {
    ctx.status = 406;
    ctx.body = 'The file is too large';
    return;
  }
  debug(`Receive file: ${fileName}, size: ${fileSize / 1024} kB`);
  const stream = fs.createWriteStream(path.join(
    __dirname, '/public/' + uuid.v4() + fileName.substr(fileName.lastIndexOf('.'))
  ));
  reader.pipe(stream);
  debug('uploading %s -> %s', file.name, stream.path);
  ctx.body = `You've uploaded the file: ${fileName}`;
}

module.exports = {
  route,
};
